CIPs/0022.md

```diff
@@ -0,0 +1,127 @@
+
+The encoding is currently implemented as:
+
+```
+encode_to_point(
+    Blake2Xs(
+        BHHash(
+            encode_integer_8_le(counter) || 
+            encode_integer_16_le(epoch number) || 
+            encode_integer_32_le(maximum_non_signers) || 
+            validator set
+        )
+    )
+)
+```
+
+where `counter` is the first number where `encode_to_point` succeeds starting from `0` and at most `255`, `validator_set` is the concatenation of the results of `encode_public_key_le` on each validator public key.
+
+We propose to change it to the following. We first define the parameters `MAX_VALIDATORS = 150` and `SEC_PARAM = 128`. We then specify the new encoding.
+
+```
+encode_to_point(
+    Blake2Xs(
+        encode_integer_16_le(epoch number) || 
+        BHHash(
+            encode_integer_8_le(counter) || 
+            bytes_le_to_bits_le(truncate(epoch_hash, SEC_PARAM/8)) ||
+            bytes_le_to_bits_le(truncate(parent_hash, SEC_PARAM/8)) ||
+            encode_integer_32_le(maximum_non_signers) || 
+            pad(validator set, MAX_VALIDATORS, G1_GENERATOR)
+        )
+    )
+)
+```
+
+where `counter` is the first number where `encode_to_point` succeeds starting from `0` and at most `255`, `validator_set` is the concatenation of the results of `encode_public_key_le` on each validator public key, `epoch_hash` is the hash of the current epoch block header, `parent_hash` is the hash of the previous epoch block header (17280 blocks in the past) and
+
+`G1_GENERATOR = (81937999373150964239938255573465948239988671502647976594219695644855304257327692006745978603320413799295628339695, 241266749859715473739788878240585681733927191168601896383759122102112907357779751001206799952863815012735208165030)`
+
+
+## Rationale
+
+A single Plumo proof can support a fixed amount of validators and epochs. We believe that 150 validators and 130 epochs (~= 130 days) will satisfy Celo's needs for the next 2 years.
+
+Moving the epoch number outside the `BHHash` allows for an efficient on-chain double-signing proof - this proof doesn't need the information inside the `BHHash`, it's just required to show there are two aggregate signatures by a super majority on the same epoch number and distinct opaque outputs of `BHHash`.
+
+Adding the current and parent block header hashes prevents the "future subcommittee" attack since it adds an unpredictable value that the attacker can't predict in advance and therefore can't pre-sign, and establishes a chain of epoch headers. Specifically, the header contains the state hash, which in turn contains the result of the Celo randomness beacon. We take only `SEC_PARAM/8` bytes, since, assuming the bytes are sufficiently unpredictable, an attacker would need to pre-sign around `2^128` different messages in order to succeed.
```

Comment:
Some more details about the future subcommittee attack would be good.  For one thing, I had a hard time finding information about it by googling it.  But, also, at first I thought since the entropy includes two header hashes, we only need to take `SEC_PARAM/16` bytes from each.  Then I thought maybe not, because the attack could perhaps be done during one epoch, so the parent hash could be know.  But then it wasn't clear to me why there should be `SEC_PARAM/8` bytes from the parent hash.  Is it to ease the SNARK by having the two hashes be the same length?  Regardless, I guess my point is just that more detail would be good on these points, both about the attack (can be a link to details elsewhere), and about the requirements for the chain of epoch headers that are behind the choice to include `SEC_PARAM/8` for the parent hash too.